target/arduino-due.hpp

```diff
@@ -218,159 +233,619 @@ namespace hwstl {
-
-                if (main_clock_frequency == MainClockFrequency) {
-                    return 0;
-                } else if (main_clock_frequency == MainClockFrequency / 2) {
-                    return 1;
-                } else if (main_clock_frequency == MainClockFrequency / 4) {
-                    return 2;
-                } else if (main_clock_frequency == MainClockFrequency / 8) {
-                    return 3;
-                } else if (main_clock_frequency == MainClockFrequency / 16) {
-                    return 4;
-                } else if (main_clock_frequency == MainClockFrequency / 32) {
-                    return 5;
-                } else if (main_clock_frequency == MainClockFrequency / 64) {
-                    return 6;
-                } else if (main_clock_frequency == MainClockFrequency / 3) {
-                    return 7;
-                } else {
-                    return -1; // Invalid clock frequency
-                }
-            }
+        class uart_port {
+        public:
+            enum class peripheral {
+                uart,
+                usart0,
+                usart1,
+                usart2
+            };
 
-            static constexpr int32_t FromPrescalerSelectorToFrequency(uint32_t prescaler_selection) {
-                if (prescaler_selection == 0) {
-                    return MainClockFrequency;
-                } else if (prescaler_selection == 1) {
-                    return MainClockFrequency / 2;
-                } else if (prescaler_selection == 2) {
-                    return MainClockFrequency / 4;
-                } else if (prescaler_selection == 3) {
-                    return MainClockFrequency / 8;
-                } else if (prescaler_selection == 4) {
-                    return MainClockFrequency / 16;
-                } else if (prescaler_selection == 5) {
-                    return MainClockFrequency / 32;
-                } else if (prescaler_selection == 6) {
-                    return MainClockFrequency / 64;
-                } else if (prescaler_selection == 7) {
-                    return MainClockFrequency / 3;
-                } else {
-                    return -1; // Invalid prescaler
+            enum class type {
+                uart,
+                usart
+            };
+
+            class peripheral_info {
+            public:
+                type m_type;
+                peripheral m_peripheral;
+            };
+
+            static constexpr peripheral_info uart { type::uart, peripheral::uart };
+            static constexpr peripheral_info usart0 { type::usart, peripheral::usart0 };
+            static constexpr peripheral_info usart1 { type::usart, peripheral::usart1 };
+            static constexpr peripheral_info usart2 { type::usart, peripheral::usart2 };
+
+            constexpr type get_type_of_peripheral(peripheral p) {
+                if (p == peripheral::uart) {
+                    return type::uart;
+                } else if (
```

Comment:
May be switched to an `else` statement. The SAM3X8E has only one UART, so any other serials should always be USART.

Reply:
Since this statement is evaluated during compilation, no unexpected side-effects should be allowed. Calling get_type_of_peripheral(<errornous-peripheral>) will cause a compilation error: `error: constexpr call flows off the end of the function`

Reply:
Fair enough.